### PR TITLE
Support for running tests in db isolation mode

### PR DESCRIPTION
**Pull request by potiuk**
Introduce a possibility to run tests in "DB isolation" mode.

Such tests will require an internal-api component to be running alongside the tests and some specific configuration that will be shared via environment variables.

This one adds support for breeze testing tests command and for the interactive breeze shell command.

It already works for most tests in dag_processor - with few caveats to fix.

It is not yet wired to the CI but it should be a very easy special test case to add - once we get the tests "green".

The gist of it is to make sure that when the
`RUN_TESTS_WITH_DATABASE_ISOLATION` environment variable is set to `true`, there is a special TracebackSesssionForTests created that will allow test code to run DB operations (setip/teardown and initialization in tests while it will not allow "airflow" code to create a DB session (it will raise exception) and all the airflow code should use internal_api in this case.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/settings.py [hunk 1]
```diff
@@ -281,6 +281,59 @@ def remove(*args, **kwargs):
...
 
+AIRFLOW_PATH_FOR_TESTS = os.path.dirname(os.path.dirname(__file__))
+AIRFLOW_TESTS_PATH_FOR_TESTS = os.path.join(AIRFLOW_PATH_FOR_TESTS, "tests")
+
+
+class TracebackSessionForTests:
+    """
+    Session that throws error when you try to create a session outside of the test code.
+
+    When we run our tests in "db isolation" mode we except that "airflow" code will never create
```
Comment (aritra24) [inline airflow/settings.py]: Expect*

File: contributing-docs/testing/unit_tests.rst [hunk 2]
```diff
@@ -738,6 +738,39 @@ You can also use fixture to create object that needs database just like this.
         conn = request.getfixturevalue(conn)
         ...
 
+Running tests with Database isolation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Running tests with DB isolation is a special case of tess that require ``internal-api`` component to be
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst]: Tests*

File: contributing-docs/testing/unit_tests.rst [hunk 3]
```diff
@@ -738,6 +738,39 @@ You can also use fixture to create object that needs database just like this.
...
+started in order to execute the tests. Only selected tests can be run with the database isolation
+(TODO: add the list) - they are simulating running untrusted components (dag file processor, triggerer,
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst:746]: Is the todo something missed or something to be added in a follow-up later?

**Review by aritra24 — APPROVED** [review]
Overall looks good to me, a few nits

File: airflow/settings.py [hunk 4]
```diff
@@ -281,6 +281,59 @@ def remove(*args, **kwargs):
...
+        is "airflow/tests" or "regular airflow".
+
+        :meta: private
+        :return: True if the object was created from test code, False otherwise.
+        """
+        for tb in self.traceback:
+            if tb.filename.startswith(AIRFLOW_PATH_FOR_TESTS):
+                if tb.filename.startswith(AIRFLOW_TESTS_PATH_FOR_TESTS):
+                    return True
+                return False
```
Comment (vincbeck) [inline airflow/settings.py]:
```suggestion
                return tb.filename.startswith(AIRFLOW_TESTS_PATH_FOR_TESTS):
```

File: contributing-docs/testing/unit_tests.rst (hunk 3 above)
Comment (potiuk) [reply contributing-docs/testing/unit_tests.rst:746]: Follow up later - we have no full list yet.

File: airflow/settings.py (hunk 4 above)
Comment (potiuk) [reply airflow/settings.py]: I am not so sure if it's better. I kind of liked the explicitness of the if -> return. But I added some comments and shortened the constants that are probably better.

**Comment by potiuk** [comment]
Updated.

**Review by vincbeck — APPROVED** [review]

**Pull request merged**